Editline enhancements and misc cleanups

zer0-droids opened the pull request:
This series makes many changes to the ec_editline API, and introduces a new example called "simple-editline", which aims to be a quick-start example for someone that wants to use libecoli. The series also makes some "menage".

zer0-droids committed:
editline: add api documentation

Add doxygen-like API comments into ec_editline.h.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: add missing const to function params

Add missing const to ec_editline_print_helps() and ec_editline_get_node()
parameters.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: check grammar node is sh_lex

Completion and parsing are only supposed to work if the top level node is of
type sh_lex. Enforce this.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: fix set prompt after set prompt_esc

Currently, ec_editline_set_prompt() does not call el_set(EL_PROMPT) to set the
prompt in libedit. This is done once at init. On the other hand,
ec_editline_set_prompt_esc() calls el_set(EL_PROMPT_ESC), which overrides the
libedit prompt.

Set the prompt in both functions.

As a reminder, see the el_set manual:

> el_set()
>     Set editline parameters. op determines which parameter to set, and each operation has its own parameter list.
>
>     The following values for op are supported, along with the required argument list:
>
> EL_PROMPT , char *(*f)(EditLine *)
>    Define prompt printing function as f, which is to return a string that contains the prompt.

Link: https://linux.die.net/man/3/el_set
Fixes: 6da7e430ae47 ("editline: allow escape sequences in prompt")
Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: remove useless full_line param from ec_editline_get_helps()

It is not needed to pass both line and full_line, since only one is used by the
function. Keep "line", which becomes optional: it can be retrieved using
ec_editline_curline() most of the time.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: remove useless full_line variable in ec_editline_complete()

full_line is not used, remove it.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: remove useless param from ec_editline_get_suggestions()

full_line is not used and not relevant, remove it.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: rename get_suggestions as get_error_helps

This function shall only be used after a parsing error, let's rename it
accordingly. A later commit will introduce a *_print_error_helps().

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: rework ec_editline_get_error_helps()

Currently, the tokenization is done in a quick and dirty way: when a space is
found in the string, it is replaced by a \0 and it retries to parse/complete to
generate the error helps.

Use the ec_strvec_sh_lex_str() function to enhance the tokenization. It makes
the function more complex, but it can be expected to be more reliable.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: introduce ec_editline_print_error_helps()

Similar to what exists for ec_editline_get_helps()/ec_editline_print_helps(),
introduce ec_editline_print_error_helps() to display the helps generated by
ec_editline_get_error_helps().

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: introduce helpers to set node attributes

These 2 functions ec_editline_set_help() and ec_editline_set_callback() can be
used to add a well-known attribute to a grammar node. These attributes are used
for interactive contextual help or for specifying a callback function to execute
on successful parsing.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
examples: introduce simple-editline

This example shows how to use the ec_editline API, in the simplest way.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
string: introduce a function to quote a string

It can for instance be used to print the representation of a string.
This function will be used in the next commit.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
strvec: quote string when dumping an strvec

Currently, the strings dumped by ec_strvec_dump() are not quoted. This can lead
to ambiguous dumps. For instance, an empty vector or a vector containing one
empty string are both displayed like this: [].

Quote the strings when dumping an strvec.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
examples/parse-yaml: remove useless variable and comments

The "line" variable is not used. Remove it.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: fix error stream in ec_editline_complete()

To get the error stream, the fileno is 2.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: remove irrelevant comments

Remove the XXX comments, it pollutes the code for almost nothing.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
examples/parse-yaml: quote dumped strings

Use the recently-introduced function ec_str_quote() to protect the dumped
strings.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
remove todo.txt file

This file contains internal notes, no need to store it in the repo.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
treewide: remove duplicated semicolon

Remove duplicated semicolon.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: use an enum for init flags

Using an enum is clearer than macros, especially for documentation: flags are
grouped into one type.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: remove useless parameter from ec_editline_parse()

The sh_lex node is stored in the editline structure, no need to pass it.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
editline: rename function parameters

Rename some parameters to be more consistent, or more relevant knowing the
meaning of the parameter.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids committed:
libecoli.sh: fix node iteration

The node id can be empty. To check that a node id is valid, check that
its type instead of its id.

This fixes the parse-yaml.sh example:

> $ bash ./parse-yaml.sh -p /path/to/ecoli-parse-yaml -l /path/to/libecoli.sh
> > hello john
> ec_node1_id=''
> ec_node1_type='sh_lex'
> ec_node1_strvec_len=1
> ec_node1_str0='hello john '
> ec_node1_first_child='ec_node2'
> ec_node2_parent='ec_node1'
> ec_node2_id=''
> ec_node2_type='or'
> ec_node2_strvec_len=2
> ec_node2_str0='hello'
> ec_node2_str1='john'
> ec_node2_first_child='ec_node3'
> ec_node3_parent='ec_node2'
> ec_node3_id='hello'
> ec_node3_type='seq'
> ec_node3_strvec_len=2
> ec_node3_str0='hello'
> ec_node3_str1='john'
> ec_node3_first_child='ec_node4'
> ec_node4_parent='ec_node3'
> ec_node4_id=''
> ec_node4_type='str'
> ec_node4_strvec_len=1
> ec_node4_str0='hello'
> ec_node4_next='ec_node5'
> ec_node5_parent='ec_node3'
> ec_node5_id='name'
> ec_node5_type='or'
> ec_node5_strvec_len=1
> ec_node5_str0='john'
> ec_node5_first_child='ec_node6'
> ec_node6_parent='ec_node5'
> ec_node6_id=''
> ec_node6_type='str'
> ec_node6_strvec_len=1
> ec_node6_str0='john'
> john says hello to you!

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

rjarry commented on include/ecoli/editline.h:
```diff
@@ -36,51 +36,61 @@ struct ec_editline_help {
+ * Type of callback invoked by ec_editline_interact() on successful parsing.
  */
-#define EC_EDITLINE_DEFAULT_SIGHANDLER 0x08
+typedef int (*ec_editline_command_cb_t)(const struct ec_pnode *);
```
It looks like this typedef ended up in the wrong commit (it is in https://github.com/rjarry/libecoli/pull/17/commits/7c05b49542d5ec95b8baca353fea34753d23e45a and should probably be in https://github.com/rjarry/libecoli/pull/17/commits/22c9259b36eef65af2bd5f4979806a46c253fa96 or https://github.com/rjarry/libecoli/pull/17/commits/9129f6a4991f8cfb03bcdcce4b6a72afd03ca434).

zer0-droids replied:
yes, good catch, fixed in coming rework

rjarry commented on src/editline.c:
```diff
@@ -290,10 +290,14 @@ ec_editline_get_node(const struct ec_editline *editline)
 ec_editline_set_node(struct ec_editline *editline, const struct ec_node *node)
 {
+	if (strcmp(ec_node_get_type_name(node), "sh_lex"))
+		return -1;
```
maybe set errno to EINVAL?

zer0-droids replied:
I added it for this function. I think that the other functions of the file will need another pass (in another patchset) to check if errno is properly set.

rjarry commented on src/editline.c:
```diff
@@ -140,26 +156,17 @@ ec_editline_set_prompt(struct ec_editline *editline, const char *prompt)
 			return -1;
 	}
 
+	if (el_set(editline->el, EL_PROMPT, prompt_cb))
+		goto fail;
```
I have read the commit message several times but I fail to understand how that works.

What is the purpose of that callback?

zer0-droids replied:
From libedit manual:

> el_set()
>     Set editline parameters. op determines which parameter to set, and each operation has its own parameter list.
>
>     The following values for op are supported, along with the required argument list:
>
> EL_PROMPT , char *(*f)(EditLine *)
>    Define prompt printing function as f, which is to return a string that contains the prompt.

prompt_cb is the function that is passed to libedit to get the prompt.

I added this quote in the commit log.

rjarry commented on src/editline.c:
```diff
@@ -129,6 +129,22 @@ ec_editline_free_helps(struct ec_editline_help *helps, size_t n)
+	void *clientdata;
+
+	if (el_get(el, EL_CLIENTDATA, &clientdata))
+		return "> ";
```
is that OK to return a literal string on failure?

zer0-droids replied:
That was already the case before the commit (this function is just moved).
But this should not happen: it would mean that ec_editline forgot to set the clientdata before calling set_prompt.

rjarry reviewed (changes requested):
https://github.com/rjarry/libecoli/pull/17/commits/30a85c98e65565f46d39f6539e9cb820bdd044e2 unrelevant -> irrelevant

rjarry commented:
general remark about commit titles: I avoid ending them with punctuation characters.

```
editline: remove useless full_line param from ec_editline_get_helps()
editline: remove useless full_line variable in ec_editline_complete()
editline: remove useless parameter from ec_editline_parse()
editline: remove useless param from ec_editline_get_suggestions()
editline: rework ec_editline_get_error_helps()
editline: introduce ec_editline_print_error_helps()
editline: introduce ec_editline_interact()
editline: fix error stream in ec_editline_complete()
```

Could you remove the trailing parentheses `()`? They don't bring much to the table.

vjardin commented on include/ecoli/editline.h:
```diff
@@ -36,51 +36,61 @@ struct ec_editline_help {
+ * Flags passed at ec_editline initialization.
  */
-#define EC_EDITLINE_DISABLE_SIGNALS 0x01
+enum ec_editline_init_flags {
```
Either explicitly set a type for your enum (`enum ec_editline_init_flags : uint8_t`) or remove the leading 0 from your flags, as the default underlying type for flags may not be uint8_t.

`c23`

https://en.cppreference.com/w/c/language/enum.html

rjarry replied:
> Either explicitly set a type for your enum (enum ec_editline_init_flags : uint8_t) or remove the leading 0 from your flags, as the default underlying type for flags may not be uint8_t.

libecoli uses c99. It would be quite a big leap to jump to c23. Also, we don't really care about the enum type size. But I concur that to avoid confusion you could rewrite this as follows:

```c
enum ec_editline_init_flags {
	EC_EDITLINE_DISABLE_SIGNALS = (1 << 0),
	EC_EDITLINE_DISABLE_HISTORY = (1 << 1),
	EC_EDITLINE_DISABLE_COMPLETION = (1 << 2),
	EC_EDITLINE_DEFAULT_SIGHANDLER = (1 << 3),
};
```

zer0-droids replied:
I updated it as suggested by @rjarry

zer0-droids committed:
editline: introduce ec_editline_interact()

This function is the main helper to run an interactive command line with
libedit. It parses, completes, invoke the callback on successful parsing, and
display errors on parsing error.

Signed-off-by: Olivier Matz <zer0@droids-corp.org>
Reviewed-by: Robin Jarry <robin@jarry.cc>

zer0-droids commented:
> [30a85c9](https://github.com/rjarry/libecoli/commit/30a85c98e65565f46d39f6539e9cb820bdd044e2) unrelevant -> irrelevant

fixed

zer0-droids commented:
> general remark about commit titles: I avoid ending them with punctuation characters.
> 
> ```
> editline: remove useless full_line param from ec_editline_get_helps()
> editline: remove useless full_line variable in ec_editline_complete()
> editline: remove useless parameter from ec_editline_parse()
> editline: remove useless param from ec_editline_get_suggestions()
> editline: rework ec_editline_get_error_helps()
> editline: introduce ec_editline_print_error_helps()
> editline: introduce ec_editline_interact()
> editline: fix error stream in ec_editline_complete()
> ```
> 
> Could you remove the trailing parentheses `()`? They don't bring much to the table.

The purpose of the parenthesis is to highlight that these names refer to functions, not variables.

For instance:
editline: introduce ec_editline_interact()

Without the (), ec_editline_interact could be a variable or a type. For that reason I'll tend to keep the title as is. What do you think?

rjarry commented:
> The purpose of the parenthesis is to highlight that these names refer to functions, not variables.
> 
> For instance: editline: introduce ec_editline_interact()
> 
> Without the (), ec_editline_interact could be a variable or a type. For that reason I'll tend to keep the title as is. What do you think?

I think that in a commit title, the difference between a function, type or variable name does not matter. And usually, you can be more specific in the commit body to clarify *if* there is an ambiguity.

I am asking to remove them because in my other series (#16), I am adding commit log checks that will report an error if a commit title does not end with an alpha-numeric character. This is mainly to prevent people from writing real sentences (yes, I have seen that multiple times).

What do you think?

zer0-droids commented:
> > The purpose of the parenthesis is to highlight that these names refer to functions, not variables.
> > For instance: editline: introduce ec_editline_interact()
> > Without the (), ec_editline_interact could be a variable or a type. For that reason I'll tend to keep the title as is. What do you think?
> 
> I think that in a commit title, the difference between a function, type or variable name does not matter. And usually, you can be more specific in the commit body to clarify _if_ there is an ambiguity.
> 
> I am asking to remove them because in my other series (#16), I am adding commit log checks that will report an error if a commit title does not end with an alpha-numeric character. This is mainly to prevent people from writing real sentences (yes, I have seen that multiple times).
> 
> What do you think?

Well, running "git log --oneline | grep ')$' | wc -l" in linux kernel tree gives more than 75K results. While I understand your point about punctuation in titles, I think parenthesis deserve a special treatment.

What do you think? :D

rjarry commented:
Ok fair enough.

I'll adjust the script to allow closing parentheses.

What do you think ? :trollface:
